src/backend/common/CommandBuffer.cpp

```diff
@@ -238,9 +257,75 @@ namespace backend {
             return true;
         };
 
+        // TODO(kainino@chromium.org): Manage this state inside an object, change lambda to a method
+        RenderPassBase* currentRenderPass = nullptr;
+        FramebufferBase* currentFramebuffer = nullptr;
+        uint32_t currentSubpass = 0;
+        auto beginSubpass = [&]() -> bool {
+            auto& subpassInfo = currentRenderPass->GetSubpassInfo(currentSubpass);
+            for (auto& kv : subpassInfo.colorAttachments) {
+                if (auto* tv = currentFramebuffer->GetTextureView(kv.second)) {
+                    auto* texture = tv->GetTexture();
+                    if (!texture->IsTransitionPossible(nxt::TextureUsageBit::ColorAttachment)) {
+                        device->HandleError("Can't transition attached texture to ColorAttachment usage");
+                        return false;
+                    }
+                    if (!textureHasGuaranteedUsageBit(texture, nxt::TextureUsageBit::ColorAttachment)) {
+                        mostRecentTextureUsages.erase(texture);
+                        texturesTransitioned.insert(texture);
+                    }
+                }
+            }
+            return true;
+        };
+
         Command type;
         while(iterator.NextCommandId(&type)) {
             switch (type) {
+                case Command::AdvanceSubpass:
+                    {
+                        iterator.NextCommand<AdvanceSubpassCmd>();
+                        if (currentRenderPass == nullptr) {
+                            device->HandleError("Can't advance subpass without an active render pass");
+                            return false;
+                        }
+                        if (currentSubpass + 1 >= currentRenderPass->GetSubpassCount()) {
+                            device->HandleError("Can't advance beyond the last subpass");
+                            return false;
+                        }
+                        currentSubpass += 1;
+                        if (!beginSubpass()) {
+                            return false;
+                        }
+                    }
+                    break;
+
+                case Command::BeginRenderPass:
+                    {
+                        BeginRenderPassCmd* begin = iterator.NextCommand<BeginRenderPassCmd>();
+                        auto* renderPass = begin->renderPass.Get();
+                        auto* framebuffer = begin->framebuffer.Get();
+                        if (currentRenderPass != nullptr) {
+                            device->HandleError("A render pass is already active");
+                            return false;
+                        }
+                        if (framebuffer->GetRenderPass() != renderPass) {
+                            device->HandleError("Framebuffer is incompatible with this render pass");
+                            return false;
+                        }
+                        lastPipeline = nullptr;
```

Comment:
What the reason to reset the pipeline and bindgroups? I think there could be a use for inheriting bindgroups from outside the pipeline.

I'm not sure what the best way to go is, but this is inconsistent with endrenderpass that also reset vertex and index buffers.